### PR TITLE
feat(run-events): export the run event vocabulary and schemas as veryfront/run-events

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

New public entry point `veryfront/run-events` (decision P13 in the producer design): the run-event vocabulary and schemas, so consumers such as Studio import them instead of hand-writing them. Stacked on #4470.

- `RUN_EVENT_TYPES` (the 53 catalogued types, pinned by count and by a digest of the sorted names against the veryfront-api catalog), `RUN_EVENT_CLASSES`, and the wire-name maps (`toRunEventWireName` / `fromRunEventWireName`), cross-checked against `NATIVE_RUN_EVENTS`.
- Schemas through `defineSchema`: the typed envelope (`event_id`, `run_id`, `event_type`, `event_class`, span and turn ids, `origin_*`, `unrecoverable_fields`, `created_at`, `is_error`), the typed row (`payload` key; plus the conversation-route variant with the `event` key, tagged `// legacy: removed in Phase F`), and one payload schema getter per type (36 getters, catalog shapes; the chat decoder's inline schemas remain deliberately looser and may consume these later).
- `parseTypedRunEventRow`; a clear `MISSING_EXTENSION_ERROR` naming the `SchemaValidator` contract and the `@veryfront/ext-schema-zod` registration when no validator is registered (no fallback validator).
- Browser-safe (`BROWSER_SAFE_EXPORTS`); a bundle test pins the retained builtins to exactly `node:async_hooks`, the same residual `./chat/ag-ui` already carries.
- Docs: guide section and JSDoc on every export; API reference regenerated.

## Test plan

- [x] 46 steps across the vocabulary, envelope, payload, validator-registration and fixture-contract tests (all eight fixture samples validate in both shapes) plus the browser-safe bundle ratchet
- [x] `deno task fmt`, `lint`, `lint:style`, `lint:anti-slop`, `lint:barrel-jsdoc`, `lint:wildcard-exports`, `lint:client-bundle`, `lint:test-semantic-dispositions`, `typecheck`, `docs:api-reference:check`; `deno task test:unit` green; Bun suite green; Node suite green except the pre-existing `src/routing/api/module-loader/loader.test.ts` failure (unrelated, needs its own ticket); `lint:test-typecheck` shows only the pre-existing `src/tool/remote-mcp.test.ts` failure from #4468
- [x] Reviewed (approved; the reviewer independently reconstructed the 53-type list from the API and matched the digest)
- [ ] Codex review on the PR
- [ ] CI

https://claude.ai/code/session_01NonPHDcbWsisd2GFB68ALo

## Summary by CodeRabbit

* **New Features**
  * Added a public typed run-events API for event types, wire-name conversion, classifications, schemas, and row parsing.
  * Added validation for run-event envelopes and payloads, including support for durable and live event data.
  * Added browser-safe support for the run-events API.

* **Documentation**
  * Added API reference documentation and a guide section explaining how to read and validate typed run events.
  * Updated documentation navigation ordering.

* **Tests**
  * Added comprehensive coverage for event vocabularies, schemas, parsing, validation, integration, and browser compatibility.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Automated review: 92/100 — excellent, minor suggestions only

Clean, well-tested new public entry point with a solid cross-repo drift-detection story.

**Strengths**
- Layering is clean: `vocabulary.ts` (types/wire names) → `schema-validator.ts` (lazy, no-fallback validator resolution with an actionable error) → `envelope.ts`/`payload.ts` (schemas) → `index.ts` (barrel). Each module has a single clear job and the JSDoc explains *why*, not just *what*.
- Forward/backward compatibility is deliberately designed in: `event_type` is an open string (not the closed enum) so a reader doesn't choke on a type the API adds later, payload schemas are `.passthrough()`, and `isRunEventType`/`fromRunEventWireName` return null/false rather than throwing on unknown input.
- Genuinely good test coverage: 46 steps across vocabulary, envelope, payload, validator-registration, and an 8-sample cross-repo fixture contract test validating both the run-scoped (`payload`) and conversation-scoped (`event`) row shapes. Edge cases are covered (null `event_id` for unpersisted live frames, missing validator error message, unknown wire names, payload key passthrough).
- The producer/reader drift-prevention mechanism (pinning `RUN_EVENT_TYPES` against the API catalog by count + SHA-256 digest of sorted names, and cross-checking native types against `NATIVE_RUN_EVENTS`) is a thoughtful way to catch cross-repo contract drift without vendoring the API's code.
- Third commit shows the PR already went through one review round (re-indent fix, payload map type narrowing) and addressed it cleanly — good sign of iteration quality.

**Minor notes (non-blocking)**
- The SHA-256 digest pinned in `vocabulary.test.ts` against the veryfront-api catalog is unverifiable from this repo/review — it's inherently trust-based on the author's transcription being correct at the stated date. Worth double-checking that value was computed off the actual current API source rather than transcribed by hand, since a wrong pin would falsely "pass" a drifted list.
- Test plan checklist doesn't explicitly list `lint:module-boundaries` / `lint:extension-contracts`, which seem relevant given a new module with cross-module imports (`#veryfront/agent/ag-ui/native-run-events.ts`). Likely covered by a broader `deno task lint`/`verify` run, but worth confirming explicitly given this is a new public entry point.
- Two checklist items are still unchecked (Codex review, CI) — please make sure both come back green before merge, per the stacking note (this is stacked on #4470).

Nothing here blocks approval; this is largely ready pending the outstanding CI/Codex checks.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run-events/envelope.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ * payload failed the API's own per-type validation still arrives with its type
+ * intact and the failing field paths listed in `unrecoverable_fields` (the API
+ * reports the damage rather than dropping the row), so validating the payload
+ * strictly at this level would reject rows the contract says to keep. Use
+ * `RUN_EVENT_PAYLOAD_SCHEMAS` or a per-type getter for the narrow check.
+ */
+export const getTypedRunEventRowSchema = defineRunEventSchema((v) =>
+  getRunEventEnvelopeSchema().extend({
+    payload: v.object({ type: v.string().min(1) }).passthrough(),
+  })
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/envelope.ts:105]: Fixed in cca09877d2: both typed-row schemas (`payload` key and the conversation `event` key) now require the nested `type` to equal `event_type` via a refinement at the nested path (under `format=typed` the API always sets `payload.type`; verified against the API's typed row schema and mapper). Six tests cover agreement, disagreement and a missing nested type for both shapes.

File: docs/guides/memory-and-streaming.md [hunk 2]
```diff
@@ -305,6 +305,83 @@ export default function ChatPage() {
...
+);
+const body = await response.json() as { data: unknown[] };
+
+for (const raw of body.data) {
+  const row = parseTypedRunEventRow(raw);
+  if (!isRunEventType(row.event_type)) continue; // a type this build predates
+  const result = RUN_EVENT_PAYLOAD_SCHEMAS[row.event_type]?.().safeParse(row.payload);
+  if (result?.success) {
+    console.log(row.event_type, row.span_id, result.data);
+  }
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply docs/guides/memory-and-streaming.md]: Fixed in cca09877d2: the guide's reader example renders every row — unknown types log the envelope plus raw payload instead of `continue`, and known types without a per-type schema (the sixteen `AGENT_RUN_*` control-plane types) fall back to the API-validated raw payload; the "one payload schema per type" bullet now names that exception.

**Comment by kojiwakayama** [comment]
@codex review

File: src/run-events/envelope.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ * `event_id` is null only for a live SSE frame that was never persisted (a
+ * heartbeat, a synthetic terminal frame). Every stored row has one, and it is
+ * the cursor value to resume from.
+ */
+export const getRunEventEnvelopeSchema = defineRunEventSchema((v) =>
+  v.object({
+    event_id: v.number().int().nonnegative().nullable(),
+    event_type: v.string().min(1),
+    run_id: v.string().min(1),
+    event_class: v.enum(RUN_EVENT_CLASSES),
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/envelope.ts:35]: Fixed in 3964dd9e40: both typed-row schemas now reject a row whose `event_class` disagrees with the class the catalog assigns to its `event_type` (a per-type class map pinned against the API by digest 75a339e0…); uncatalogued types accept any class, matching the API's open `event_type` rule. Tests: agreement, disagreement, uncatalogued.

File: scripts/build/browser-safe-exports.mjs [hunk 4]
```diff
@@ -17,6 +17,13 @@ export const BROWSER_SAFE_EXPORTS = [
   "./markdown",
   "./mdx",
   "./agent/identity",
+  // The typed run event contract. Studio reads a run's event log in the
+  // browser bundle. The entry point is vocabulary and schemas only: no npm
+  // dependency, no server module, and the single Node builtin it reaches
+  // (node:async_hooks, via the contract registry) is the same one ./chat and
+  // ./chat/ag-ui already reach and is rewritten to a no-op browser polyfill.
+  // browser-safe-exports.test.ts pins that set so nothing worse creeps in.
+  "./run-events",
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/build/browser-safe-exports.mjs:26]: Accepted as a pre-existing residual rather than fixed here: the one retained builtin is `node:async_hooks`, reached through the contract registry's module-scope `AsyncLocalStorage` (`extensions/contracts.ts` → `contract-registry-internal.ts` → `platform/compat/async-context.ts`). `./chat/ag-ui` and `./chat` retain the same one today, the import rewriter maps it to a no-op browser polyfill, and `scripts/build/browser-safe-exports.test.ts` pins the retained set for `./run-events` to exactly that builtin. Removing it means making the registry's async context lazy, which touches every schema-carrying export; tracked as #4473.

File: src/run-events/index.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ * @example
+ * ```ts
+ * import {
+ *   isRunEventType,
+ *   parseTypedRunEventRow,
+ *   RUN_EVENT_PAYLOAD_SCHEMAS,
+ * } from "veryfront/run-events";
+ *
+ * const response = await fetch(`${apiUrl}/runs/${runId}/events?format=typed`, {
+ *   headers: { Authorization: `Bearer ${token}` },
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/index.ts:54]: Fixed in 8c5eb184eb: the module `@example` declares `apiUrl`, `runId` and `token` with safe placeholders so it is complete and pasteable; the generated reference is regenerated.

File: src/run-events/index.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+ * });
+ * const body = await response.json() as { data: unknown[] };
+ *
+ * for (const raw of body.data) {
+ *   const row = parseTypedRunEventRow(raw);
+ *   if (!isRunEventType(row.event_type)) continue;
+ *   const result = RUN_EVENT_PAYLOAD_SCHEMAS[row.event_type]?.().safeParse(row.payload);
+ *   if (result?.success) {
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/index.ts]: Fixed in 8c5eb184eb: the `@example` falls back to the API-validated raw payload when no per-type schema exists (the sixteen `AGENT_RUN_*` control-plane types), mirroring the corrected guide example, so no row is skipped.

**Comment by kojiwakayama** [comment]
@codex review

File: docs/guides/memory-and-streaming.md [hunk 7]
```diff
@@ -305,6 +305,95 @@ export default function ChatPage() {
...
+```ts
+import {
+  isRunEventType,
+  parseTypedRunEventRow,
+  RUN_EVENT_PAYLOAD_SCHEMAS,
+} from "veryfront/run-events";
+
+const response = await fetch(
+  `${apiUrl}/runs/${runId}/events?format=typed`,
+  { headers: { Authorization: `Bearer <API_TOKEN>` } },
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply docs/guides/memory-and-streaming.md]: Fixed in d40d8daceb: the guide example now declares `apiUrl`, `runId` and `token` with the same placeholders as the module JSDoc and uses `token` in the Authorization header.

**Comment by kojiwakayama** [comment]
@codex review

File: src/run-events/index.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * @module run-events
+ *
+ * @example
+ * ```ts
+ * import {
+ *   isRunEventType,
+ *   parseTypedRunEventRow,
+ *   RUN_EVENT_PAYLOAD_SCHEMAS,
+ * } from "veryfront/run-events";
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/index.ts:41]: Fixed in 65663a45e3: the module example now imports `register` and `createZodAdapter` and registers the `SchemaValidator` before the first parse (the guide example gets the same lines).

File: src/run-events/envelope.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+ *
+ * @param input - One row from a typed run event surface.
+ * @returns The parsed row.
+ * @throws When the row is missing an envelope field or a typed payload.
+ *
+ * @example
+ * ```ts
+ * import { parseTypedRunEventRow } from "veryfront/run-events";
+ *
+ * const response = await fetch(`${apiUrl}/runs/${runId}/events?format=typed`);
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/envelope.ts]: Fixed in 65663a45e3: the parser example declares `apiUrl`, `runId` and `token` with the same placeholders as the module example and sends the Authorization header.

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kojiwakayama** [comment]
@codex review

File: src/run-events/index.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+ * import { createZodAdapter } from "@veryfront/ext-schema-zod";
+ * import {
+ *   isRunEventType,
+ *   parseTypedRunEventRow,
+ *   RUN_EVENT_PAYLOAD_SCHEMAS,
+ * } from "veryfront/run-events";
+ *
+ * // Outside a Veryfront app, register a validator once at startup; inside
+ * // one, bootstrap already has and this line is an idempotent no-op.
+ * register("SchemaValidator", createZodAdapter());
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/index.ts]: Fixed in 464afe48cb: every example and the guidance (module docs, guide, and the getter's error message) now gate registration on `tryResolve("SchemaValidator")`, and the idempotence claim is replaced by the accurate statement that `register` replaces whatever is registered.

**Comment by kojiwakayama** [comment]
@codex review

File: src/run-events/vocabulary.ts [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+  "TOOL_CALL_CHUNK",
+  "REASONING_MESSAGE_CONTENT",
+  "REASONING_CONTENT",
+  "STATE_DELTA",
+  "ACTIVITY_DELTA",
+]);
+
+/** The event class the API reports for a stored type. */
+export function getRunEventClass(eventType: string): RunEventClass {
+  return RUN_EVENT_DELTA_TYPES.has(eventType) ? "delta" : "fact";
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/vocabulary.ts]: Fixed in 37d6015373: `getRunEventClass` returns `null` for a type outside the catalog (same convention as `fromRunEventWireName`), with JSDoc pointing callers at the row's `event_class` for such types; the per-type table derives from the catalogued-only classifier.

File: docs/api-reference/veryfront/run-events.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
+---
+title: "veryfront/run-events"
+description: "The typed run event contract the Veryfront API publishes. A run event surface read with `format=typed` returns rows that carry a span envelope and a payload named by a catalogued type. This module owns the reader's half of that contract: the type vocabulary, the AG-UI wire names, the envelope and row schemas, and one payload schema per type. Import it instead of writing the names or shapes out again. Every schema here is lazy and materializes through the registered `SchemaValidator` contract, so a consumer outside a Veryfront app must register one before the first `get*Schema()` call or `parseTypedRunEventRow`. Registration is idempotent and takes one line at startup: ```ts import { register } from \"veryfront/extensions/contracts\"; import { createZodAdapter } from \"@veryfront/ext-schema-zod\"; register(\"SchemaValidator\", createZodAdapter()); ``` Inside a Veryfront app, bootstrap does this before handlers run. This module ships no fallback validator: calling a getter with nothing registered throws an error naming the contract and this registration call."
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply docs/api-reference/veryfront/run-events.md]: Fixed in 37d6015373: the API reference page and index are regenerated from the corrected JSDoc (`deno task docs:api-reference:check` passes).

**Comment by kojiwakayama** [comment]
@codex review

File: src/run-events/index.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+ * const token = "<TOKEN>";
+ *
+ * const response = await fetch(`${apiUrl}/runs/${runId}/events?format=typed`, {
+ *   headers: { Authorization: `Bearer ${token}` },
+ * });
+ * const body = await response.json() as { data: unknown[] };
+ *
+ * for (const raw of body.data) {
+ *   const row = parseTypedRunEventRow(raw);
+ *   if (!isRunEventType(row.event_type)) continue;
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/index.ts]: Fixed in 981d319028: the module example logs the raw payload of a row whose type this build predates before continuing, as the guide does.

File: src/run-events/envelope.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * @param input - One row from a typed run event surface.
+ * @returns The parsed row.
+ * @throws When the row is missing an envelope field or a typed payload.
+ *
+ * @example
+ * ```ts
+ * import { parseTypedRunEventRow } from "veryfront/run-events";
+ *
+ * // Requires a registered `SchemaValidator`; see the module docs.
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply src/run-events/envelope.ts]: Fixed in 981d319028: the parser example carries the guarded `tryResolve`/`register` setup itself, so it runs standalone.

**Comment by kojiwakayama** [comment]
@codex review
